src/vs-picgo/index.ts

```diff
@@ -0,0 +1,192 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import * as os from 'os';
+import * as vscode from 'vscode';
+
+import PicGo from 'picgo/dist/core/PicGo';
+import { ImgInfo } from 'picgo/dist/utils/interfaces';
+
+import { promisify } from 'util';
+
+import { formatParam, formatString } from '../utils';
+
+const _ = require('lodash');
+const _db = require('lodash-id');
+_.mixin(_db);
+
+const writeFileP = promisify(fs.writeFile);
+const readFileP = promisify(fs.readFile);
+
+export interface INotice {
+  body: string;
+  text: string;
+  title: string;
+}
+
+export interface IUploadName {
+  date: string;
+  dateTime: string;
+  fileName: string;
+  extName: string;
+  mdFileName: string;
+  [key: string]: string;
+}
+
+export interface IOutputUrl {
+  uploadedName: string;
+  url: string;
+  [key: string]: string;
+}
+
+export default class VSPicgo {
+  private dataPath: string;
+  private editor: vscode.TextEditor;
+  private picgo: PicGo;
+
+  constructor(editor: vscode.TextEditor) {
+    const picgoConfigPath = vscode.workspace.getConfiguration('picgo').get<string>('configPath');
+    if (picgoConfigPath) {
+      this.picgo = new PicGo(picgoConfigPath);
+    } else {
+      this.picgo = new PicGo();
+      const picBed = vscode.workspace.getConfiguration('picgo.picBed');
+      this.picgo.setConfig({ picBed });
+    }
+    this.editor = editor;
+    this.dataPath = this.getDataPath();
+    // Before upload, we change names of the images.
+    this.addRenameListener();
+    // After upload, we use the custom output format.
+    this.addGenerateOutputListener();
+  }
+
+  addGenerateOutputListener() {
+    this.picgo.on('finished', async (ctx: PicGo) => {
+      let urlText = '';
+      const outputFormatTemplate =
+        vscode.workspace.getConfiguration('picgo').get<string>('customOutputFormat') || '![${uploadedName}](${url})';
+      try {
+        urlText = ctx.output.reduce((acc: string, cur: ImgInfo): string => {
+          return `${acc}${formatString(outputFormatTemplate, {
+            uploadedName: cur.fileName || '',
+            url: cur.imgUrl,
+          })}\n`;
+        }, '');
+        urlText = urlText.trim();
+        await this.updateLog(ctx.output);
+      } catch (err) {
+        if (err instanceof SyntaxError) {
+          vscode.window.showErrorMessage(
+            `The data file ${this.dataPath} has syntax error, ` +
+              `please fix the error by yourself or delete the data file and vs-picgo will recreate for you.`,
+          );
+        } else {
+          vscode.window.showErrorMessage(`Failed to read from data file ${this.dataPath}: ${err || ''}`);
+        }
+        return;
+      }
+      this.editor.edit(textEditor => {
+        textEditor.replace(this.editor.selection, urlText);
+        vscode.window.showInformationMessage('Upload successfully');
```

Comment:
We should keep the message format consistent, for example:
1. There is no Upper case word at the beginning, need a header: `Picgo: `
![image](https://user-images.githubusercontent.com/24741764/56474974-508fc100-64b4-11e9-82c5-599cc2fad8c1.png)
2. `Upload` -> `Uploaded`, since we have uploaded the message.

Reply:
I have made them consistent:
![image](https://user-images.githubusercontent.com/24741764/56827578-ca97bf80-6891-11e9-9c63-afbd567e2064.png)
![image](https://user-images.githubusercontent.com/24741764/56827472-85738d80-6891-11e9-9223-b48d68787773.png)
![image](https://user-images.githubusercontent.com/24741764/56827537-b5229580-6891-11e9-9e13-53e846ad9243.png)